### ISSUE TITLE
incorrect padding when stride > 2?

### ISSUE THREAD

**Issue by victor-shepardson** [acids-ircam/cached_conv#10]
In `get_padding`, I think the padding should be reduced by `stride-1` -- otherwise the last few elements of the input are discarded (they only affect the next frame, via the cache). I traced an extra block of latency in RAVE models trained with `--config causal` to this.

```python
import cached_conv as cc

r = 4
for d in range(1,r+1):
    conv = cc.CachedConv1d(
        1, 1,
        kernel_size=2 * r,
        stride=r,
        padding=cc.get_padding(2 * r, r, mode='causal'),
        bias=False
    )

    x = torch.zeros(1, 1, 8)
    x[:,:,-d] = 1000
    print(conv(x))
```
outputs:
```
tensor([[[0., 0.]]], grad_fn=<ConvolutionBackward0>)
tensor([[[0., 0.]]], grad_fn=<ConvolutionBackward0>)
tensor([[[0., 0.]]], grad_fn=<ConvolutionBackward0>)
tensor([[[  0.0000, -26.7172]]], grad_fn=<ConvolutionBackward0>)
```

the logic here: https://github.com/acids-ircam/cached_conv/blob/beb8a8ab4977458ae36c1c15577952d0783662cf/cached_conv/convs.py#L34C11-L34C11 fixes this in the `stride=2, kernel=4` case, but there's still extra padding with `stride=4, kernel=8`.

I suggest changing the line `p = (kernel_size - 1) * dilation + 1` to `p = (kernel_size - 1) * dilation + 1 - (stride-1)`

I tried re-exporting a causal RAVE model after this fix, and it removed 40ms of latency:
<img width="1345" alt="Screenshot 2023-10-20 at 13 44 33" src="https://github.com/acids-ircam/cached_conv/assets/4522484/3ef60cd8-072c-4124-ae29-6a7ceb98a25c">

But I'm not 100% certain it's correct for every combination of stride/dilation/mode

**Comment by victor-shepardson** [comment]
so, re-exporting a RAVE model with the above change does reduce latency. But *training* a new model after the change brings it back. I think this is because the vanilla `ConvTranspose1d` still uses future padding during training. 

There seems to be no way to ask pytorch for asymmetric padding in transposed convolutions. Between this and the general weirdness of padding for transposed convolutions, it's not straightforward to configure a causal transposed conv via the existing `get_padding` API. Presumably this is why the use of CachedConvTranspose in RAVE doesn't do that already (https://github.com/acids-ircam/RAVE/blob/0e7a3089532657ff04f076abc18887a20b0cc82c/rave/blocks.py#L652) 

Here's my solution for allowing causal `ConvTranspose1d` which matches `CachedConvTranspose1d`:
https://github.com/victor-shepardson/cached_conv/commit/95479c31ad673cb7ac484da0bf0ca6610d7f3014 . It requires a second flag for configuring causal / non causal. It should be backward compatible though, supplying causal=False will just result in the current extra block of latency.
